Bottom: 5750b540a14138aa263260dce70c3da7a7e21d42
Top:    b8cf1a1c660e5c8cd9420cbe1cb9ada5c5b68d01
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-06-01 16:59:00 +0800

tun: avoid compute hashing on tx

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index e16487c..63a9368 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1332,7 +1332,10 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	skb_reset_network_header(skb);
 	skb_probe_transport_header(skb, 0);
 
-	rxhash = skb_get_hash(skb);
+#ifdef CONFIG_RPS
+	if (tun->numqueues == 1 && static_key_false(&rps_needed))
+		rxhash = skb_get_hash(skb);
+#endif
 	netif_rx_ni(skb);
 
 	stats = get_cpu_ptr(tun->pcpu_stats);
@@ -1342,7 +1345,10 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	u64_stats_update_end(&stats->syncp);
 	put_cpu_ptr(stats);
 
-	tun_flow_update(tun, rxhash, tfile);
+#ifdef CONFIG_RPS
+	if (tun->numqueues == 1 && static_key_false(&rps_needed))
+		tun_flow_update(tun, rxhash, tfile);
+#endif
 	return total_len;
 }
